tt_metal/jit_build/build.cpp

```diff
@@ -491,16 +494,55 @@ void JitBuildState::elf_to_hex8(const string& log_file, const string& out_dir) c
     }
 }
 
-void JitBuildState::hex8_to_hex32(const string& log_file, const string& out_dir) const
-{
-    string cmd;
-    cmd = "cd " + out_dir + " && ";
-    cmd += env_.hex8tohex32_ + this->target_name_ + ".hex.tmp " + this->target_name_ + ".hex";
-
-    log_debug(tt::LogBuildKernels, "    hex8tohex32 cmd: {}", cmd);
-    if (!tt::utils::run_command(cmd, log_file, false)) {
-        build_failure(this->target_name_, "hex8tohex32.py", cmd, log_file);
+void hex8_to_hex32(const string& log_file, const string& out_dir, const string& target_name){
+    ZoneScoped;
+    auto write_data = [](std::ofstream& outf, std::vector<long>& data, long& ptr){
+        if (!data.empty()) {
+            outf << "@" << std::setfill('0') << std::setw(8) << std::hex << (ptr >> 2) << "\n";
+            while (data.size() % 4 != 0) {
+                data.push_back(0);
+            }
```

Comment:
At the very least, make a helper function to pad since this logic is used throughout this PR.

But even better would be to do a linear scan to figure out how big your vector should be, and then do one allocation to construct your vector (with all 0s). Then, just overwrite the non-pad values with actual data.

Reply:
linear scan noticeably slows down runtime on longer test runs, python_api unit testing run below: 

<img width="763" alt="image" src="https://github.com/tenstorrent-metal/tt-metal/assets/157417475/89d16b26-0ae1-46e0-b621-a410c20ac15a">

<img width="769" alt="image" src="https://github.com/tenstorrent-metal/tt-metal/assets/157417475/24cc62c5-11e7-4355-949e-ec576c5fcf5f">

Reply:
helper function was added in most recent commit, pls take a look when you have time